### ISSUE TITLE
Cyclomatic complexity sniffer doesn't provide classes to refactor

### ISSUE THREAD

**Issue by bpolaszek** [nunomaduro/phpinsights#422]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| Library version  | 1.14.0

Hello there,

Here's an example output of my phpinsights:

```bash
# ...
• [Complexity] Having `classes` with more than 10 cyclomatic complexity is prohibited - Consider refactoring:

• [Architecture] Normal classes are forbidden. Classes must be final or abstract:
# ...
```

The console shows nothing after "consider refactoring:" (I expect a list of incriminated classes). Is it normal?

Thanks!
Ben

**Comment by Jibbarth** [comment]
Hi @bpolaszek :wave: 

You are right, it should list classes to refactor.

Can you provide your `phpinsights.php` configuration file ? 
Did you reproduce on `dev-master` ? 

I suspect that files that should be listed are excluded later, whereas the piece of code that say "we have an issue" don't filter on it.
See https://github.com/nunomaduro/phpinsights/blob/v1.14.0/src/Domain/Insights/CyclomaticComplexityIsHigh.php (L17 vs L50)

**Comment by bpolaszek** [comment]
Hey @Jibbarth  👋

Following your comments:
- Upgrading to `dev-master` has no effect
- Changing `maxComplexity` has no effect on the score
- Adding `NunoMaduro\PhpInsights\Domain\Insights\CyclomaticComplexityIsHigh::class` in the `remove` section has absolutely no effect. 

<details><summary>Here's my phpinsights.php:</summary>

```php
<?php

declare(strict_types=1);

use NunoMaduro\PhpInsights\Domain\Insights\CyclomaticComplexityIsHigh;
use NunoMaduro\PhpInsights\Domain\Insights\ForbiddenNormalClasses;
use NunoMaduro\PhpInsights\Domain\Insights\ForbiddenTraits;
use NunoMaduro\PhpInsights\Domain\Sniffs\ForbiddenSetterSniff;
use ObjectCalisthenics\Sniffs\Files\FunctionLengthSniff;
use ObjectCalisthenics\Sniffs\Metrics\MethodPerClassLimitSniff;
use ObjectCalisthenics\Sniffs\NamingConventions\ElementNameMinimalLengthSniff;
use PHP_CodeSniffer\Standards\Generic\Sniffs\Files\LineLengthSniff;
use PHP_CodeSniffer\Standards\Generic\Sniffs\Formatting\SpaceAfterNotSniff;
use SlevomatCodingStandard\Sniffs\Classes\SuperfluousAbstractClassNamingSniff;
use SlevomatCodingStandard\Sniffs\Classes\SuperfluousExceptionNamingSniff;
use SlevomatCodingStandard\Sniffs\Classes\SuperfluousInterfaceNamingSniff;
use SlevomatCodingStandard\Sniffs\Classes\SuperfluousTraitNamingSniff;
use SlevomatCodingStandard\Sniffs\Commenting\DocCommentSpacingSniff;
use SlevomatCodingStandard\Sniffs\ControlStructures\AssignmentInConditionSniff;
use SlevomatCodingStandard\Sniffs\ControlStructures\DisallowShortTernaryOperatorSniff;
use SlevomatCodingStandard\Sniffs\ControlStructures\DisallowYodaComparisonSniff;
use SlevomatCodingStandard\Sniffs\Functions\StaticClosureSniff;
use SlevomatCodingStandard\Sniffs\TypeHints\DisallowArrayTypeHintSyntaxSniff;

return [
    /*
    |--------------------------------------------------------------------------
    | Default Preset
    |--------------------------------------------------------------------------
    |
    | This option controls the default preset that will be used by PHP Insights
    | to make your code reliable, simple, and clean. However, you can always
    | adjust the `Metrics` and `Insights` below in this configuration file.
    |
    | Supported: "default", "laravel", "symfony", "magento2", "drupal"
    |
    */

    'preset' => 'symfony',

    /*
    |--------------------------------------------------------------------------
    | IDE
    |--------------------------------------------------------------------------
    |
    | This options allow to add hyperlinks in your terminal to quickly open
    | files in your favorite IDE while browsing your PhpInsights report.
    |
    | Supported: "textmate", "macvim", "emacs", "sublime", "phpstorm",
    | "atom", "vscode".
    |
    | If you have another IDE that is not in this list but which provide an
    | url-handler, you could fill this config with a pattern like this:
    |
    | myide://open?url=file://%f&line=%l
    |
    */

    'ide' => 'phpstorm',

    /*
    |--------------------------------------------------------------------------
    | Configuration
    |--------------------------------------------------------------------------
    |
    | Here you may adjust all the various `Insights` that will be used by PHP
    | Insights. You can either add, remove or configure `Insights`. Keep in
    | mind, that all added `Insights` must belong to a specific `Metric`.
    |
    */

    'exclude' => [
        'migrations',
        'deploy.php',
    ],

    'add' => [
        //  ExampleMetric::class => [
        //      ExampleInsight::class,
        //  ]
    ],

    'remove' => [
        DisallowArrayTypeHintSyntaxSniff::class,
        ElementNameMinimalLengthSniff::class,
        AssignmentInConditionSniff::class,
        DisallowYodaComparisonSniff::class,
        DocCommentSpacingSniff::class,
        SpaceAfterNotSniff::class,
        SuperfluousAbstractClassNamingSniff::class,
        SuperfluousExceptionNamingSniff::class,
        SuperfluousInterfaceNamingSniff::class,
        SuperfluousTraitNamingSniff::class,
        StaticClosureSniff::class,
        ForbiddenTraits::class,
        DisallowShortTernaryOperatorSniff::class,
    ],

    'config' => [
        ForbiddenSetterSniff::class => [
            'exclude' => [
                'src/Entity', // Entities must have setters
            ],
        ],
        ForbiddenNormalClasses::class => [
            'exclude' => [
                'src/Entity', // Entities cannot be final because of Doctrine's proxy pattern
            ],
        ],
        LineLengthSniff::class => [
            'lineLimit' => 120,
            'absoluteLineLimit' => 160,
        ],
        MethodPerClassLimitSniff::class => [
            'exclude' => [
                'src/Entity', // Entities can have lots of methods (because of setters)
            ],
            'maxCount' => 15,
        ],
        CyclomaticComplexityIsHigh::class => [
            'maxComplexity' => 10,
        ],
        FunctionLengthSniff::class => [
            'maxLength' => 40,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Requirements
    |--------------------------------------------------------------------------
    |
    | Here you may define a level you want to reach per `Insights` category.
    | When a score is lower than the minimum level defined, then an error
    | code will be returned. This is optional and individually defined.
    |
    */

    'requirements' => [
        'min-quality' => 80,
        'min-complexity' => 80,
        'min-architecture' => 80,
        'min-style' => 80,
        'disable-security-check' => false,
    ],
];

```
</details>

Thank you! 🙏 
Ben

**Comment by Jibbarth** [comment]
@bpolaszek Thanks for your response,

> Changing `maxComplexity` has no effect on the score

Yes this is an other issue :sweat: See #367 


> Adding `NunoMaduro\PhpInsights\Domain\Insights\CyclomaticComplexityIsHigh::class` in the `remove` section has absolutely no effect.

No effect at all ? Or it remove the following line ?
```
• [Complexity] Having `classes` with more than 10 cyclomatic complexity is prohibited - Consider refactoring:
```

To clarify, the class `CyclomaticComplexityIsHigh` is used to report this line above, and "normally" list incriminated classes. 

The score is based on the collector that inspect all files, and for now, we cannot change the way this score is calculated by changing any config. https://github.com/nunomaduro/phpinsights/blob/master/src/Domain/Results.php#L49-L59


I'm afraid that I'm not able to reproduce your initial problem. Could you help us on this bug ?

I think the method `getDetails()` of class `CyclomaticComplexityIsHigh` is responsible of the bug. 
Maybe could you add some `dump` (or breakpoint) inside to check how the array is fill ?

**Comment by bpolaszek** [comment]
Sure. Dumping [$complexityLimit](https://github.com/nunomaduro/phpinsights/blob/master/src/Domain/Insights/CyclomaticComplexityIsHigh.php#L38), [$classesComplexity](https://github.com/nunomaduro/phpinsights/blob/master/src/Domain/Insights/CyclomaticComplexityIsHigh.php#L47) and the [returned value](https://github.com/nunomaduro/phpinsights/blob/master/src/Domain/Insights/CyclomaticComplexityIsHigh.php#L51-L53), here's the output:
```php
^ 10
^ array:1 [
  "Dto/Membership/MembershipInputDataTransformer.php" => 14
]
^ array:1 [
  0 => NunoMaduro\PhpInsights\Domain\Details^ {#42816
    -original: null
    -file: "Dto/Membership/MembershipInputDataTransformer.php"
    -line: null
    -function: null
    -message: "14 cyclomatic complexity"
    -diff: null
  }
]
```

**Comment by 50bhan** [comment]
@bpolaszek If you add `NunoMaduro\PhpInsights\Domain\Insights\CyclomaticComplexityIsHigh::class` in the `remove` section of `phpinsights.php` file, it will work correctly right now:

![Screenshot from 2020-08-09 17-38-10](https://user-images.githubusercontent.com/36772278/89732938-4a39a580-da67-11ea-88ce-ee794281f0f7.png)

P.S: This issue fixed through #420 and it's only available on `dev-master` at this moment.

**Comment by bpolaszek** [comment]
Sorry to reopen, but unfortunately this is still unsolved to me ☹️ 

After upgrading to `dev-master`:
- If Cyclomatic Complexity has the default `maxComplexity`, complexity score is 81% and classes appear (I have some classes between 5-9 cyclomatic complexity).
- If Cyclomatic Complexity has `maxComplexity` set to `10`, classes don't appear anymore (none of them has 10+ CC) but score is still `81%`, meaning the next class I add with less than 10 CC will break my CI (threshold to 80) although all of them respecting the `maxComplexity` setting.
